Sources/StreamChat/APIClient/ChatRemoteNotificationHandler.swift

```diff
@@ -0,0 +1,199 @@
+        guard let payload = content.userInfo["stream"], let dict = payload as? [String: String] else {
+            throw ChatPushNotificationError.invalidUserInfo("missing stream key or not a [string:string] dict")
+        }
+
+        guard let type = dict["type"] else {
+            throw ChatPushNotificationError.invalidUserInfo("missing stream.type key")
+        }
+
+        eventType = EventType(rawValue: type)
+
+        if let cid = dict["cid"] {
+            self.cid = try? ChannelId(cid: cid)
+        } else {
+            cid = nil
+        }
+
+        if EventType.messageNew.rawValue == type, let id = dict["id"] {
+            messageId = MessageId(id)
+        } else {
+            messageId = nil
+        }
+    }
+}
+
+public class ChatRemoteNotificationHandler {
+    var client: ChatClient
+    var content: UNNotificationContent
+    let chatCategoryIdentifiers: Set<String> = ["stream.chat", "MESSAGE_NEW"]
+    let syncCooldown: TimeInterval = 6.0
+    let database: DatabaseContainer
+
+    public init(client: ChatClient, content: UNNotificationContent) {
+        self.client = client
+        self.content = content
+        database = client.databaseContainer
+    }
+
+    private func obtainLastSyncDate(completion: @escaping (Date?) -> Void) {
+        database.backgroundReadOnlyContext.perform {
+            completion(self.database.backgroundReadOnlyContext.currentUser?.lastReceivedEventDate)
+        }
+    }
+
+    private func bumpLastSyncData(lastReceivedEventDate: Date, completion: @escaping () -> Void) {
+        database.write { session in
+            session.currentUser?.lastReceivedEventDate = lastReceivedEventDate
+            completion()
+        }
+    }
+
+    private func shouldSyncWithAPI(since: Date, now: Date = .init()) -> Bool {
+        now.timeIntervalSince(since) > syncCooldown
+    }
+
+    private func syncChannel(completion: @escaping () -> Void) {
+        guard client.config.isLocalStorageEnabled else {
+            completion()
+            return
+        }
+
+        obtainLastSyncDate { lastSyncAt in
+            guard let lastSyncAt = lastSyncAt, self.shouldSyncWithAPI(since: lastSyncAt) else {
+                completion()
+                return
+            }
+
+            let request = ChannelDTO.allChannelsFetchRequest
+            request.fetchLimit = 1000
+            request.propertiesToFetch = ["cid"]
+
+            guard let results = try? self.database.backgroundReadOnlyContext.fetch(request) else {
+                completion()
+                return
+            }
+
+            let cids = results.map(\.cid).compactMap { try? ChannelId(cid: $0) }
+            guard !cids.isEmpty else {
+                completion()
+                return
+            }
+
+            let endpoint: Endpoint<MissingEventsPayload> = .missingEvents(
+                since: lastSyncAt,
+                cids: cids
+            )
+
+            self.client.apiClient.request(endpoint: endpoint) {
```

Comment:
We're doing an API request inside `backgroundReadOnlyContext`'s queue... We call the completion block of `obtainLastSyncDate` in this queue and it continues as so. We should switch queues.

Reply:
sorry not sure what you mean

Reply:
ideally we did everything sync since this code is meant to run in a service extension. any idea?

Reply:
@b-onc ping on this, afaict this does not matter given the environment the code runs (ext). let me know :)